packages/webdriverio/src/commands/browser/switchFrame.ts

```diff
@@ -0,0 +1,217 @@
+            const isDesiredFrame = await context(ctx)
+            if (isDesiredFrame) {
+                return context
+            }
+        }
+
+        sessionContext.setCurrentContext(currentContext)
+        throw new Error('Could not find the desired frame')
+    }
+
+    throw new Error(
+        `Invalid type for context parameter: ${typeof context}, expected one of number, string or null. ` +
+        'Check out our docs: https://webdriver.io/docs/api/browser/switchToFrame.html'
+    )
+}
+
+function switchToFrameHelper (browser: WebdriverIO.Browser, context: string) {
+    const sessionContext = getContextManager(browser)
+    sessionContext.setCurrentContext(context)
+}
+
+async function switchToFrameUsingElement (browser: WebdriverIO.Browser, element: WebdriverIO.Element) {
+    let frameSrc = await element.getAttribute('src')
+    if (!frameSrc) {
+        throw new Error(`Frame with index ${context} has no src attribute`)
+    }
+
+    if (!frameSrc.startsWith('http')) {
+        /**
+         * use the browser to access `URL.parse` as it is not available in Node.js until v22
+         */
+        frameSrc = await browser.execute((urlPath: string) => (
+            URL.parse(urlPath, window.location.href)?.href
+        ), frameSrc) || frameSrc
+    }
+
+    const tree = await browser.browsingContextGetTree({})
+    const urlContext = findContext(frameSrc, tree.contexts, byUrl)
+    if (!urlContext) {
+        throw new Error(`Frame with url "${frameSrc}" not found`)
+    }
+
+    await switchToFrameHelper(browser, urlContext)
+    return urlContext
+}
+
+function byUrl (context: local.BrowsingContextInfo, url: string) {
+    return context.url === url
+}
+
+function byUrlContaining (context: local.BrowsingContextInfo, url: string) {
+    return context.url.includes(url)
+}
+
+function byContextId (context: local.BrowsingContextInfo, contextId: string) {
+    return context.context === contextId
+}
+
+function findContext (
+    url: string,
+    contexts: local.BrowsingContextInfoList,
+    matcher: typeof byUrl | typeof byUrlContaining
+): string | undefined {
+    for (const context of contexts) {
+        if (matcher(context, url)) {
+            return context.context
+        }
+
+        if (context.children && context.children.length > 0) {
+            const result = findContext(url, context.children, matcher)
+            if (result) {
+                return result
+            }
+        }
+    }
+
+    return undefined
+}
+
+/**
+ * Given a certain `BrowsingContextInfoList` and a desired `contextId`, this helper finds
+ * the `BrowsingContextInfo` that matches the `contextId`
+ * @param tree  a context tree returned from `browsingContextGetTree`
+ * @param contextId  a desired context id
+ * @returns the `BrowsingContextInfo` that matches the `contextId` or `undefined` if not found
+ */
+function findContextById (tree: local.BrowsingContextInfoList, contextId: string): local.BrowsingContextInfo | undefined {
```

Comment:
same function as above, could potentially be merged

Reply:
Good call!